### PR TITLE
feat(frontend): add `LogicalMultijoin`

### PR DESCRIPTION
**Pull request by jon-chuang**
## What's changed and what's your intention?
Adds `LogicalMultiJoin` PlanTreeNode. 

In a follow up PR (in progress; passes e2e tests reliably but missing unit tests) we will add optimizer rules for converting a `LogicalMultiJoin` into a left-deep join tree with heuristic reordering to avoid cross joins (and prioritize e.g. joins with primary keys in join columns).

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/2425
https://github.com/singularity-data/risingwave/issues/1866

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+impl PlanTreeNode for LogicalMultiJoin {
+    fn inputs(&self) -> smallvec::SmallVec<[crate::optimizer::PlanRef; 2]> {
+        let mut vec = smallvec::SmallVec::new();
+        vec.extend(self.inputs.clone().into_iter());
+        vec
+    }
+
+    fn clone_with_inputs(&self, inputs: &[crate::optimizer::PlanRef]) -> crate::optimizer::PlanRef {
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:119]: Maybe this should be left as todo!()

File: src/frontend/src/optimizer/rule/multijoin_join.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        let multi_join = LogicalMultiJoin::from_join(&join_1.into()).unwrap();
+        for (input, schema) in
+            multi_join
+                .inputs()
+                .iter()
+                .zip_eq(vec![mid.schema(), left.schema(), right.schema()])
+        {
+            assert_eq!(input.schema(), schema);
+        }
+    }
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/rule/multijoin_join.rs:123]: Should tests that the conjunctions have been correctly rewritten.

**Comment by fuyufjh** [comment]
AFAIK, #1907 and #2425 are designed for two different purposes. #2079 

- #1907 is for multi-way delta join rewrite. Currently, the delta join only supports 2-way join, but we are working in progress to move the plan fragmenter from meta node to frontend, after that, it's not necessary to introduce a multi-join operator. You may talk to @skyzh and @st1page for details.
- #2425 is for join reordering. To implement an heuristic join reordering algorithm, a common practice is to converting all inner joins to a single multi join operator and do reordering inside of it. In my opinion, perhaps we need an RFC/proposal of join reordering first. Let's have a discussion next week? @XieJiann @st1page

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        if let Some(multi_join) = left.as_logical_multi_join() {
+            inputs.extend(multi_join.inputs());
+            conjunctions.extend(multi_join.on().clone());
+        } else {
+            inputs.push(left.clone());
+        }
+        if let Some(multi_join) = right.as_logical_multi_join() {
+            inputs.extend(multi_join.inputs());
+            let right_on = multi_join.on().clone();
+            let mut mapping = ColIndexMapping::with_shift_offset(
```
Comment (lmatz) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:70]: Could you add some comments about this mapping?

I thought for the `right_on`, the `input ref` in it should get shifted by `left_col_num` to the right?
`(0...right_col_num) -> (left_col_num...left_col_num + right_col_num)`

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_multi_join.rs:70]: Yes, basically. But here we are taking the inverse of the o2r mapping, I can change it to yours to be more clear.

**Comment by jon-chuang** [comment]
Yes, agreed on the differences. This PR is meant more to solve the second issue. The plannode is just a placeholder and is only meant to last the lifetime of a few optimization passes. It should not be propagated beyond the frontend for the time being.

I will write up the approach I chose for the heuristic ordering in an RFC.

File: src/frontend/src/optimizer/plan_node/logical_multi_join.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+impl ToBatch for LogicalMultiJoin {
+    fn to_batch(&self) -> Result<PlanRef> {
+        todo!()
+    }
+}
+
+impl ColPrunable for LogicalMultiJoin {
+    fn prune_col(&self, _required_cols: &[usize]) -> PlanRef {
+        todo!()
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_multi_join.rs:129]: I think Todo should be replaced by panic!(), We never want to implement these methods as LogicalMultiJoin is just a placeholder node with a temporary lifetime.

**Comment by fuyufjh** [comment]
By the way, @XieJiann mentioned he was working on this. Did you had any communication? 

https://github.com/singularity-data/risingwave/issues/2425#issuecomment-1123128527

**Comment by jon-chuang** [comment]
Unfortunately not. I was enthusiastic to try out the second part, but it was blocking on the first part, and after some study, I think its not a trivial issue for a new contributor, so I decided to get started to get the ball rolling.

My apologies @XieJiann, if you had some ongoing work on it. However, I hope you can help with some of the subtasks and the RFC for heuristic join ordering https://github.com/singularity-data/risingwave/issues/2530 if you have come up with any thoughts or solutions!

**Comment by skyzh** [comment]
I believe this representation is pretty okay even for multi-way delta join rewrite. 🤪 As far as I can see, it's easy make use of this logical plan to create a delta join plan.

Regarding join reordering, I think it's okay to have a function called `.reorder(new_idx: &[usize])` on this to create a new plan node based on the new join order. What do you think? cc @fuyufjh

**Comment by fuyufjh** [comment]
> I believe this representation is pretty okay even for multi-way delta join rewrite. 🤪 As far as I can see, it's easy make use of this logical plan to create a delta join plan.

Great!

> Regarding join reordering, I think it's okay to have a function called `.reorder(new_idx: &[usize])` on this to create a new plan node based on the new join order. What do you think? cc @fuyufjh

Agree, the join reordering rule (or else) should manipulate on the LogicalMultiJoin node

**Review by skyzh — APPROVED** [review]
LGTM from my side, and maybe other members can give some more comments.

**Review by lmatz — APPROVED** [review]
LGTM

**Pull request merged**